src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <assert.h>
+#include <sstream>
+
+#include "tracer.hpp"
+
+using namespace std;
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    thread_local std::shared_ptr<CTracer> tracer = nullptr;
+    std::shared_ptr<CTracer> CreateTracer()
+    {
+        tracer = std::make_shared<CTracer>();
+        return tracer;
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer's ISumulator implementation
+    //------------------------------------------------------------------------------------------------------------------
+    IQuantumGateSet* CTracer::AsQuantumGateSet()
+    {
+        return nullptr;
+    }
+    IDiagnostics* CTracer::AsDiagnostics()
+    {
+        return nullptr;
+    }
+    Qubit CTracer::AllocateQubit()
+    {
+        size_t qubit = qubits.size();
+        qubits.push_back({});
+        return reinterpret_cast<Qubit>(qubit);
+    }
+    void CTracer::ReleaseQubit(Qubit /*qubit*/)
+    {
+        // nothing for now
+    }
+    std::string CTracer::QubitToString(Qubit q)
+    {
+        size_t qubitIndex = reinterpret_cast<size_t>(q);
+        const QubitState& qstate = this->UseQubit(q);
+
+        stringstream str(qubitIndex);
+        str << " last used in layer " << qstate.layer << "(pending zero ops: " << qstate.pendingZeroOps.size() << ")";
+        return str.str();
+    }
+    void CTracer::ReleaseResult(Result /*result*/)
+    {
+        // nothing to do, we don't allocate results on measurement
+    }
+    // Although the tracer should never compare results or get their values, it still has to implement UseZero and
+    // UseOne methods as they are invoked by the QIR initialization.
+    Result CTracer::UseZero()
+    {
+        return reinterpret_cast<Result>(INVALID);
+    }
+    Result CTracer::UseOne()
+    {
+        return reinterpret_cast<Result>(INVALID);
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::CreateNewLayer
+    //------------------------------------------------------------------------------------------------------------------
+    LayerId CTracer::CreateNewLayer(Duration opDuration)

```

Comment:
Maybe rename `opDuration` to `minLayerDuration`?

Reply:
Renamed to `minRequiredDuration`.